WMF Framework/Event Platform Client/EPCNetworkManager.swift

```diff
@@ -0,0 +1,135 @@
+     */
+    private var outputBuffer: [(url: URL, body: NSDictionary)] = []
+    
+    init(session: Session = Session.shared) {
+        self.session = session
+    }
+
+    func schedulePost(url: URL, body: NSDictionary) {
+        appendPostToOutputBuffer((url: url, body: body))
+    }
+    
+    func httpDownload(url: URL, completion: @escaping (Data?) -> Void) {
+        httpDownload(url: url, retriesRemaining: 5, attemptDelay: 2, completion: completion)
+    }
+
+    /**
+     * Download data over HTTP with capacity for retries
+     * - Parameters:
+     *   - url: where to request data from
+     *   - retriesRemaining: maximum number of retries allowed for this download
+     *     operation
+     *   - attemptDelay: time between each retry
+     *   - completion: what to do with the downloaded data
+     *
+     * The operation will be performed up to `retriesRemaining + 1` times.
+     */
+    private func httpDownload(url: URL, retriesRemaining: Int, attemptDelay: TimeInterval, completion: @escaping (Data?) -> Void) {
+
+        if retriesRemaining < 0 {
+            completion(nil)
+            return
+        }
+        
+        let task = session.dataTask(with: url) { (data, response, error) in
+            
+            let failureBlock = {
+                dispatchOnMainQueueAfterDelayInSeconds(attemptDelay) {
+                    self.httpDownload(url: url, retriesRemaining: retriesRemaining - 1, attemptDelay: attemptDelay, completion: completion)
+                }
+            }
+            
+            guard let httpResponse = response as? HTTPURLResponse,
+                let data = data else {
+                    DDLogWarn("EPCNetworkManager: server did not respond adequately, will try \(url.absoluteString) again")
+                failureBlock()
+                return
+            }
+            
+            guard httpResponse.statusCode == 200 else {
+                DDLogWarn("EPCNetworkManager: HTTP status of response was \(httpResponse.statusCode), will try \(url.absoluteString) again")
+                failureBlock()
+                return
+            }
+            
+            guard error == nil else {
+                failureBlock()
+                return
+            }
+            
+            completion(data)
+        }
+        task?.resume()
+    }
+    
+    func httpTryPost() {
+        var queuedEvent: (url: URL, body: NSDictionary)?
+        while !outputBufferIsEmpty() {
+            queuedEvent = removeOutputBufferAtIndex(0)
+            if let queuedEvent = queuedEvent {
+                self.submit(url: queuedEvent.url, payload: queuedEvent.body)
+            }
+        }
+    }
+
+    private func submit(url: URL, payload: Any? = nil) {
+        let request = session.request(with: url, method: .post, bodyParameters: payload, bodyEncoding: .json)
+        let task = session.dataTask(with: request, completionHandler: { (_, response, error) in
```

Comment:
One last thing before I forget - it would be good to ask the system to extend the background execution time before calling this. This [documentation](https://developer.apple.com/documentation/uikit/app_and_environment/scenes/preparing_your_ui_to_run_in_the_background/extending_your_app_s_background_execution_time) has a good write-up on it, though note the network call is synchronous in their sample code. I think you should be able to end the background task in the completionHandler here since this is asynchronous.

Reply:
Wow, cool, yeah I think we should definitely do this.

Reply:
```
'shared' is unavailable in application extensions for iOS: Use view controller based solutions where appropriate instead.
```

Let's table this for the revision?

Reply:
Did you try using [performExpiringActivity](https://developer.apple.com/documentation/foundation/processinfo/1617030-performexpiringactivity) instead as suggested at the bottom of the docs that Toni linked?

Reply:
@mdholloway I totally didn't see that! Good eyes. Alright, maybe I can work Toni's suggestion in after all.

Reply:
Another thought I had is, move these EPC protocol conformances to `PeriodicWorker` and `BackgroundFetcher` into their own file, then move that file out of WMF and into the Wikipedia project (uncheck the WMF target and check the (Wikipedia + Staging + Experimental + User Testing targets, you also might want to move the file location out of the WMF area in the project directory). So the code to do periodic work lives in the app, but the rest of the library still lives in WMF Framework. Then you'd have to rework postAllScheduled() to take a completion block, and add the `beginBackgroundTask` code around it. @mdholloway 's solution sounds potentially faster and easier though, I'm just not familiar with it (we don't have any places in the codebase where we use `performExpiringActivity`).

```
//MARK: PeriodicWorker
extension EPC: PeriodicWorker {
    public func doPeriodicWork(_ completion: @escaping () -> Void) {
        self.postAllScheduled()
    }
}

//MARK: BackgroundFetcher
extension EPC: BackgroundFetcher {
    public func performBackgroundFetch(_ completion: @escaping (UIBackgroundFetchResult) -> Void) {
        doPeriodicWork {
            completion(.noData)
        }
    }
}
```

Reply:
Captured in wishlist for v2 ([T261987](https://phabricator.wikimedia.org/T261987))